File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-4.cpp
// left half is going to be greater than the right half
// we can tell if we're in the left half by mid >= left
// if we're in the left side of the array then we can see if
// our target is within the bounds of our left side.
// if target < left or target > mid, that means we should search -->

class Solution 
{
public:
    int search(vector<int>& nums, int target) 
    {
        int nums_len = static_cast<int>(nums.size());
        int left = 0, right = nums_len - 1;

        while (left <= right)
        {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) { return mid; }

            // now we have to determine which half we're in
            if (nums[mid] >= nums[left])
            {
                if (target < nums[left] || target > nums[mid])
                {
                    left = mid + 1;
                }
                else
                {
                    right = mid - 1;
                }
            }
            // this is if we're in the right half
            else
            {
                if (target > nums[right] || target < nums[mid])
                {
                    right = mid - 1;
                }
                else
                {
                    left = mid + 1;
                }
            }
        }
        return -1;
    }
};

File: Data Structures & Algorithms/search-2d-matrix/submission-4.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        
        top, bottom = 0, len(matrix) - 1

        while top <= bottom:
            mid = (top + bottom) // 2

            if matrix[mid][0] <= target <= matrix[mid][-1]:
                print("here")
                # Normal Binary Search
                left, right = 0, len(matrix[0]) - 1

                while left <= right:
                    mid2 = (left + right) // 2

                    if matrix[mid][mid2] == target:
                        return True
                    elif matrix[mid][mid2] > target:
                        right = mid2 - 1
                    else:
                        left = mid2 + 1
                return False

            elif matrix[mid][-1] < target:
                top = mid + 1
            else:
                bottom = mid - 1

        return False